# **Basic Regex Exercise**

**Class Exercise Part 02a**

FAA tower closures

A list of FAA tower closures has been copied from a [PDF](https://www.faa.gov/news/media/fct_closed.pdf) into the file faa.txt.


In [90]:
# read the file into a single string
with open('/content/faa.txt') as f:
    data = f.read()

In [91]:
# check the number of characters
len(data)

5574

In [92]:
# examine the first 500 characters
data[0:500]

'FAA Contract Tower Closure List\n(149 FCTs)\n3-22-2013\nLOC\nID Facility Name City State\nDHN DOTHAN RGNL DOTHAN AL\nTCL TUSCALOOSA RGNL TUSCALOOSA AL\nFYV DRAKE FIELD FAYETTEVILLE AR\nTXK TEXARKANA RGNL-WEBB FIELD TEXARKANA AR\nGEU GLENDALE MUNI GLENDALE AZ\nGYR PHOENIX GOODYEAR GOODYEAR AZ\nIFP LAUGHLIN/BULLHEAD INTL BULLHEAD CITY AZ\nRYN RYAN FIELD TUCSON AZ\nFUL FULLERTON MUNI FULLERTON CA\nMER CASTLE ATWATER CA\nOXR OXNARD OXNARD CA\nRAL RIVERSIDE MUNI RIVERSIDE CA\nRNM RAMONA RAMONA CA\nSAC SACRAMENTO EXECU'

In [93]:
# examine the last 500 characters
data[-500:]

' YAKIMA WA\nCWA CENTRAL WISCONSIN MOSINEE WI\nEAU CHIPPEWA VALLEY RGNL EAU CLAIRE WI\nENW KENOSHA RGNL KENOSHA WI\nPage 3 of 4\nFAA Contract Tower Closure List\n(149 FCTs)\n3-22-2013\nLOC\nID Facility Name City State\nJVL SOUTHERN WISCONSIN RGNL JANESVILLE WI\nLSE LA CROSSE MUNI LA CROSSE WI\nMWC LAWRENCE J TIMMERMAN MILWAUKEE WI\nOSH WITTMAN RGNL OSHKOSH WI\nUES WAUKESHA COUNTY WAUKESHA WI\nHLG WHEELING OHIO CO WHEELING WV\nLWB GREENBRIER VALLEY LEWISBURG WV\nPKB MID-OHIO VALLEY RGNL PARKERSBURG WV\nPage 4 of 4\n'

Your assignment is to **create a list of tuples** containing the **tower IDs** and the **states** they are located in.

Here is the **expected output**:

faa = [('DHN', 'AL'), ('TCL', 'AL'), ..., ('PKB', 'WV')]

In [94]:
import re

In [96]:
# method 1
faa  = re.findall(r'\n(\w{3}) .+ (\w{2})',data)


In [97]:
print(faa)

[('DHN', 'AL'), ('TCL', 'AL'), ('FYV', 'AR'), ('TXK', 'AR'), ('GEU', 'AZ'), ('GYR', 'AZ'), ('IFP', 'AZ'), ('RYN', 'AZ'), ('FUL', 'CA'), ('MER', 'CA'), ('OXR', 'CA'), ('RAL', 'CA'), ('RNM', 'CA'), ('SAC', 'CA'), ('SDM', 'CA'), ('SNS', 'CA'), ('VCV', 'CA'), ('WHP', 'CA'), ('WJF', 'CA'), ('BDR', 'CT'), ('DXR', 'CT'), ('GON', 'CT'), ('HFD', 'CT'), ('HVN', 'CT'), ('OXC', 'CT'), ('APF', 'FL'), ('BCT', 'FL'), ('EVB', 'FL'), ('FMY', 'FL'), ('HWO', 'FL'), ('LAL', 'FL'), ('LEE', 'FL'), ('OCF', 'FL'), ('OMN', 'FL'), ('PGD', 'FL'), ('SGJ', 'FL'), ('SPG', 'FL'), ('SUA', 'FL'), ('TIX', 'FL'), ('ABY', 'GA'), ('AHN', 'GA'), ('LZU', 'GA'), ('MCN', 'GA'), ('RYY', 'GA'), ('DBQ', 'IA'), ('IDA', 'ID'), ('LWS', 'ID'), ('FAA', 'Li'), ('PIH', 'ID'), ('SUN', 'ID'), ('ALN', 'IL'), ('BMI', 'IL'), ('DEC', 'IL'), ('MDH', 'IL'), ('UGN', 'IL'), ('BAK', 'IN'), ('GYY', 'IN'), ('HUT', 'KS'), ('IXD', 'KS'), ('MHK', 'KS'), ('OJC', 'KS'), ('TOP', 'KS'), ('OWB', 'KY'), ('PAH', 'KY'), ('DTN', 'LA'), ('BVY', 'MA'), ('EWB', '

Use regular expressions to extract the **number of closures** listed in the second line of the file (149), and then use an **assertion** to check that the number of closures is equal to the length of the faa list.

In [99]:
# method 1 
# extracting the exact data

#hypothesis_closure = re.search(r'(\d+\s)',data)
#num_closures = int(hypothesis_closure.group()) # Works, but not using

num_closures = int(re.search(r'\((\d+) FCTs\)', data).group(1))

#Length Check
assert(len(faa) == num_closures) # getting a mismatch, 152 for faa

False

In [100]:
print(num_closures)

149


**Class Exercise Part 02b**: Stack Overflow reputation

I have downloaded a Stack Overflow reputation history into the file reputation.txt.

We are only interested in the lines that begin with two dashes, such as:

> -- 2012-08-30 rep +5    = 6

That line can be interpreted as follows: "On 2012-08-30, reputation increased by 5, bringing reputation total to 6."

Your assignment is to create a list of tuples containing only these dated entries, including the date, reputation change (regardless of whether it is positive/negative/zero), and running total.

Here is the **expected output**:

> rep = [('2012-08-30', '+5', '6'), ('2012-12-11', '+10', '16'), ...,  ('2015-10-14', '-1', '317')]

In [101]:
# read the file into a single string
with open('/content/reputation.txt') as f:
    data = f.read()

In [102]:
# examine the first 500 characters
data[:500]

'total votes: 36\n 2  12201376 (5)\n-- 2012-08-30 rep +5    = 6         \n 2  13822612 (10)\n-- 2012-12-11 rep +10   = 16        \n 2  13822612 (10)\n-- 2013-03-20 rep +10   = 26        \n-- 2013-12-05 rep 0     = 26        \n-- 2014-01-25 rep 0     = 26        \n 16  7141669 (2)\n-- 2014-03-19 rep +2    = 28        \n 1  12202249 (2)\n-- 2014-05-11 rep +2    = 30        \n 16 23599806 (2)\n 2  23597220 (10)\n-- 2014-05-12 rep +12   = 42        \n 2  13822612 (10)\n-- 2014-06-12 rep +10   = 52        \n 2  2359722'

In [103]:
rep = re.findall(r'-\s(\d+-\d+-\d+)\srep\s(.+?)\s+=\s(\d+)',data)
rep

[('2012-08-30', '+5', '6'),
 ('2012-12-11', '+10', '16'),
 ('2013-03-20', '+10', '26'),
 ('2013-12-05', '0', '26'),
 ('2014-01-25', '0', '26'),
 ('2014-03-19', '+2', '28'),
 ('2014-05-11', '+2', '30'),
 ('2014-05-12', '+12', '42'),
 ('2014-06-12', '+10', '52'),
 ('2014-06-26', '+10', '62'),
 ('2014-07-05', '0', '62'),
 ('2014-09-02', '0', '62'),
 ('2014-09-03', '+10', '72'),
 ('2014-10-28', '0', '72'),
 ('2014-11-14', '+10', '82'),
 ('2014-11-18', '+2', '84'),
 ('2014-12-08', '+2', '86'),
 ('2014-12-09', '+10', '96'),
 ('2014-12-12', '+2', '98'),
 ('2014-12-24', '+10', '108'),
 ('2015-02-03', '0', '108'),
 ('2015-02-20', '+10', '118'),
 ('2015-03-28', '+10', '128'),
 ('2015-04-26', '+10', '138'),
 ('2015-05-05', '+10', '148'),
 ('2015-05-26', '+10', '158'),
 ('2015-05-27', '+20', '178'),
 ('2015-06-09', '0', '178'),
 ('2015-07-03', '+10', '188'),
 ('2015-07-06', '0', '188'),
 ('2015-07-22', '+110', '298'),
 ('2015-08-21', '+10', '308'),
 ('2015-09-07', '+10', '318'),
 ('2015-10-14', '-

Convert this list of tuples into a **pandas DataFrame**. It should have appropriate column names, and the second and third columns, for example: 'change' and 'running_total', should be of type integer (rather than string/object).

In [105]:
# convert the list of tuples into a DataFrame
import pandas as pd
col_names = ['date', 'change', 'running_total']
repframe = pd.DataFrame(rep, columns = col_names)
repframe = repframe.astype({'change':'int32','running_total':'int32'}) #formatted on a second line, had an issue trying to combine ines
repframe

date  change  running_total
0   2012-08-30       5              6
1   2012-12-11      10             16
2   2013-03-20      10             26
3   2013-12-05       0             26
4   2014-01-25       0             26
5   2014-03-19       2             28
6   2014-05-11       2             30
7   2014-05-12      12             42
8   2014-06-12      10             52
9   2014-06-26      10             62
10  2014-07-05       0             62
11  2014-09-02       0             62
12  2014-09-03      10             72
13  2014-10-28       0             72
14  2014-11-14      10             82
15  2014-11-18       2             84
16  2014-12-08       2             86
17  2014-12-09      10             96
18  2014-12-12       2             98
19  2014-12-24      10            108
20  2015-02-03       0            108
21  2015-02-20      10            118
22  2015-03-28      10            128
23  2015-04-26      10            138
24  2015-05-05      10            148
25  2015-05-26      10            158
26  2015-05-27      20            178
27  2015-06-09       0            178
28  2015-07-03      10            188
29  2015-07-06       0            188
30  2015-07-22     110            298
31  2015-08-21      10            308
32  2015-09-07      10            318
33  2015-10-14      -1            317